Jetpack Connect: Optimize site lists requests

roccotripaldi opened the pull request:
Jetpack connect's authorization currently fetches the entire site list before AND after connecting.

This PR attempts to only fetch the single site we are connecting instead.

To test:
- get this PR going at calypso.localhost:3000
- run through the different connection scenarios at p2: P7rd6c-Me-p2 and ensure there are no regressions
- if you are using a wordpress.com account with a lot of sites, for example an automattic account, it should be much faster 🎉

lezama commented on client/state/jetpack-connect/actions.js:
```diff
@@ -358,9 +359,13 @@ export default {
+					site: omit( data, '_headers' )
+				} );
+				dispatch( {
+					type: SITE_REQUEST_SUCCESS,
```
do we need to dispatch this one?

sirreal commented on client/signup/jetpack-connect/site-card.jsx:
```diff
@@ -47,7 +48,7 @@ class SiteCard extends Component {
 		return (
 			<CompactCard className="jetpack-connect__site">
-				<QuerySites allSites />
+				<QuerySites siteId={ parseInt( client_id ) } />
```
`siteId` is `cliend_id`? Is that right?

tyxla replied:
Is there a case where `client_id` can be returned from the API as a non-numeric value? That could query site with ID `0` or `1`, so let's confirm that this is not a possible case.

That also applies for when we perform the network request in the `authorize()` action.

tyxla commented on client/state/sites/test/reducer.js:
```diff
@@ -226,6 +226,23 @@ describe( 'reducer', () => {
+			const state = items( original, {
+				type: SITE_REQUEST_FAILURE,
+				siteId: 2916284,
+				status: { }
```
`SITE_REQUEST_FAILURE` actions don't contain a `status`, but they do contain a `error`.

tyxla commented on client/state/sites/reducer.js:
```diff
@@ -105,6 +105,9 @@ export function items( state = {}, action ) {
 		case SITE_DELETE_RECEIVE:
 			return omit( state, action.siteId );
 
+		case SITE_REQUEST_FAILURE:
```
It feels a little odd to remove a site if it fails to be loaded. What if we have it in state already, but the request fails because of timeout or a temporary server overload? Do we want to remove that site in that case?

In case you prefer to keep this behavior, you can combine the similar actions that omit a site:

```
case SITE_DELETE_RECEIVE:
case SITE_REQUEST_FAILURE:
case JETPACK_DISCONNECT_RECEIVE:
	return omit( state, action.siteId );
```

roccotripaldi replied:
This is excellent feedback! It felt wrong the way I was handling removal, but i couldn't think of an alternative. Thanks for the suggestion!

tyxla commented on client/state/sites/test/reducer.js:
```diff
@@ -226,6 +226,23 @@ describe( 'reducer', () => {
 			} );
 		} );
 
+		it( 'should remove an unreachable site', () => {
```
Let's also add a test case that covers removing a site that is not present (returning original state in that case).

tyxla commented on client/lib/sites-list/index.js:
```diff
@@ -26,9 +25,6 @@ module.exports = function() {
 						_sites.sync( action.data );
 					}
 					break;
-				case JETPACK_CONNECT_AUTHORIZE_RECEIVE_SITE_LIST:
```
❤️

tyxla commented on client/state/jetpack-connect/actions.js:
```diff
@@ -348,22 +348,22 @@ export default {
 				dispatch( {
-					type: JETPACK_CONNECT_AUTHORIZE_RECEIVE_SITE_LIST,
-					data: data
+					type: SITE_REQUEST_SUCCESS,
```
To be consistent, we should also make sure to dispatch a `SITE_DELETE_FAILURE` if the request fails. Maybe we could directly inject `requestSite()` instead of duplicating its functionality?

roccotripaldi replied:
Great suggestion. I understand what you are saying, but I'm not sure how to implement.

roccotripaldi commented:
Thanks so much @tyxla - I pushed a change that covers most of your feedback.
A few things I'd like to address tomorrow:

- create a new query component explicitly for JPC, so that we only remove an unreachable site within the context of JPC
- address `client_id` being 1 or 0

roccotripaldi commented:
Ok! This is ready for an other review / testing.

roccotripaldi commented:
@lezama and I were testing this today and noticed a strange behavior:

- land on the JPC authorization page
- calypso by default issues a requests to `me/sites`
- finish up connection real quick
- land on plans page
- the request to `me/sites` completes and wipes the new site from the site's list

I think when we are in the Jetpack Connect system, Calypso does not need to make requests to `me/sites` as JPC handles site list management rather explicitly. I'll see if I can solve that in some way.

roccotripaldi commented:
> One way to handle this is to not render the logged-in Master Bar on Jetpack Connect routes, because I believe that is where the me/sites query originates.

Nope. This is not the case, the query originates in `layout/index.jsx` - so there is no way around it. But, maybe thunk middleware can help us. Researching now...

roccotripaldi commented:
Well, there you have it. Thunk FTW!!

sirreal commented:
> Well, there you have it. Thunk FTW!!

Does this mean that the strange behavior you observed is now fixed?

tyxla commented:
Yup @roccotripaldi, it's expected to query all sites in layout, that gives a whole lot of options throughout Calypso, without worrying if sites are already there.

Can you please clarify what is the problem when the request to `me/sites` completes and wipes the new site that was fetched previously? I'd expect that there wouldn't be anything to wipe, because the site from `/me/sites/` would be the same as the one that we previously requested? I'm asking mainly because I'm in doubt about that freezing strategy - it feels like a code smell to me.

roccotripaldi commented:
I don't think it is a code smell, it's just that JPC requires more surgical handling of site's list. The behavior I'm proposing only needs to apply JPC process.

Here is the strange behavior again.

1. We land on the authorization page, the new site is not yet in the site's list, and calypso issues a request to me/sites
2. We quickly click approve, the connection completes, and we add only the new site to site's list
3. We land on the plans page, and the query from step 1 completes
4. Because that query was issued before the new site existed, the site's list is updated without the new site.

Perhaps I do not need to do a freeze, but I'm totally jazzed that redux gives us that possibility!

> Does this mean that the strange behavior you observed is now fixed?

Yep, it is currently fixed. But I can try to find an other solution. I cannot imagine any solutions that do not use thunk middleware.  For example, maybe instead of freezing the site's list, I can dispatch `UPDATE_SITES` instead of `RECEIVE_SITES`.

roccotripaldi commented:
Ok! So I've removed all mentions of 'freeze site list' and opted for what I feel is a more elegant approach. We now have a new branch in jetpackConnect state: jetpackConnectSitesList, which will contain a flag indicating that a new site has been added. I also added a co-responding selector, hasNewlyConnectedSite.

Now the sites list will dispatch SITES_UPDATE instead of SITES_RECEIVE when a newly connected site is present. This will ensure that the entire list is not overwritten. 

I'm feeling pretty good about it. Now I need folks to test away!

tyxla commented on client/state/jetpack-connect/reducer.js:
```diff
@@ -126,6 +127,18 @@ export function jetpackConnectSite( state = {}, action ) {
 	return state;
 }
 
+export function jetpackConnectSitesList( state = {}, action ) {
```
Do we need this under a new reducer tree? The `newSite` would make perfect sense in `jetpackConnectAuthorize` IMO.

roccotripaldi replied:
Yes. I'd like to keep it in a new reducer tree. When we we dispatch JETPACK_CONNECT_COMPLETE_FLOW, i'd like jetpackConnectAuthorize to be an empty state. However, because of the requirements described above, in some cases we'll need to know when a new site is connected. That's why I added this to a separate branch.

tyxla commented on client/state/jetpack-connect/test/actions.js:
```diff
@@ -261,26 +260,46 @@ describe( 'actions', () => {
 			} );
 
-			it( 'should dispatch sites receive action when request completes', () => {
+			it( 'should dispatch authorize receive site list action when request completes', () => {
```
Nitpicking here: it's a `site` receive, not a `site list` receive action, right?

tyxla commented on client/state/jetpack-connect/actions.js:
```diff
@@ -348,23 +351,24 @@ export default {
 				userFactory().fetch();
-				return wpcom.me().sites( { site_visibility: 'all', include_domain_only: true } );
+				dispatch( {
+					type: SITE_REQUEST,
```
We're dispatching a `SITE_REQUEST`, indicating that a site request has started, and a `SITE_REQUEST_SUCCESS` upon successful request. However, if the request fails, we're not dispatching a `SITE_REQUEST_FAILURE`, so if the request fails for some reason, we'll not get notified that the request has finished; and a subsequent request to the same site might not be handled properly because, we'll incorrectly assume that a site request is currently in process.

tyxla commented on client/state/jetpack-connect/selectors.js:
```diff
@@ -146,6 +146,19 @@ const getGlobalSelectedPlan = function( state ) {
+
+const hasNewlyConnectedSite = function( state ) {
+	const jetpackConnectSitesList = get( state, [ 'jetpackConnect', 'jetpackConnectSitesList' ] );
+	return ( jetpackConnectSitesList && jetpackConnectSitesList.newSite );
```
Do we need the parens?

tyxla commented on client/state/jetpack-connect/selectors.js:
```diff
@@ -146,6 +146,19 @@ const getGlobalSelectedPlan = function( state ) {
+	if ( authorizationData.queryObject && authorizationData.queryObject.client_id ) {
+		return parseInt( authorizationData.queryObject.client_id );
+	}
+	return false;
```
I'd expect a `get*` selector to return a `null` and not a boolean `false` if data is not found.

tyxla commented on client/state/jetpack-connect/test/selectors.js:
```diff
@@ -662,4 +664,55 @@ describe( 'selectors', () => {
+	describe( '#hasNewlyConnectedSite()', () => {
+		it( 'should be undefined with an empty state', () => {
+			const state = {};
+			expect( hasNewlyConnectedSite( state ) ).to.be.undefined;
```
The name of the selector (`has*`) implies that it returns a boolean value. How do you feel about returning a `false` instead of `undefined` in both of the cases where it now returns an `undefined`?

sirreal commented on client/state/jetpack-connect/actions.js:
```diff
@@ -348,23 +351,24 @@ export default {
+				return wpcom.site( siteId ).get();
 			} )
 			.then( ( data ) => {
 				tracksEvent( dispatch, 'calypso_jpc_auth_sitesrefresh', {
```
Should this event name be updated?

roccotripaldi replied:
I'm going to leave it as is, just in case we are using the event in any funnels.

tyxla commented on client/state/sites/actions.js:
```diff
@@ -107,9 +114,10 @@ export function requestSites() {
  * a site.
  *
  * @param  {Number}   siteId Site ID
+ * @param  {Boolean}  removeSiteOnFailure Remove the site from list if request fails
```
Nitpicking: Let's keep the separate JSDoc sections aligned together vertically, in this case - `Site ID` with `Remove the site ...`.

tyxla commented on client/state/sites/reducer.js:
```diff
@@ -103,8 +103,6 @@ export function items( state = {}, action ) {
 			}, initialNextState );
```
There is a comment here that suggests us to remove the `if ( SITES_UPDATE === action.type && ! memo[ site.ID ] ) {` check when `sites-list` is removed. However, that can damage our use case of this action, right? We might want to remove that comment just in case, cc @aduth for a second opinion on that.

aduth replied:
The thinking with the condition is that we'd only previously tracked the currently selected site in state. We'd want to allow updates to be received and applied for those selected site(s), but not all sites until all sites are contained within state. Otherwise we'd be duplicating data between sites list and state, and much of the state data would go unused (since "all sites" data continued to consume exclusively from sites-list).

With recent changes to sites-list, it may be the case that we're now tracking all sites in state? If so, the initial condition could/should be dropped altogether.

tyxla replied:
Thanks for chiming in @aduth.

So yeah, the thing is that in our particular case here with Jetpack Connect a site that we're not tracking CAN appear after we've started the request to fetch all sites. So we'll have to keep this check for good, and this only confirms that we'll have that ditch comment from there.

aduth replied:
I've not caught up to the specific requirements, but speaking generally, I think we'd want sites in states to be kept in sync with updates as they're received.

tyxla replied:
@aduth of course. This is really an edge case though, let me clarify what happens and why we need to keep that update behavior:

* We load `/jetpack/connect`.
* The `<QuerySites />` in layout.jsx issues a request to `/me/sites`.
* We connect a site, and add it to the sites state so we can use it.
* The `/me/sites` request is old, and does not contain that site yet; so when that request finishes, that site is removed from the state.

In that case we want to keep that newly connected site instead of removing it.

tyxla commented on client/state/jetpack-connect/reducer.js:
```diff
@@ -253,7 +266,7 @@ export function jetpackAuthAttempts( state = {}, action ) {
 		case DESERIALIZE:
 		case SERIALIZE:
-			state;
+			return state;
```
This could have been caught if we had tests for it. Let's add some.

sirreal replied:
Test are good, and I'm for adding tests, but they wouldn't catch because this is followed by an identical return. These lines are actually redundant. Do we want them to be here for some sort of clarity, or should they be removed?

sirreal replied:
Test are good, and I'm for adding tests, but they wouldn't catch this because it's followed by an identical return. These lines are actually redundant. Do we want them to be here for some sort of clarity, or should they be removed?

tyxla replied:
Good 👁  @sirreal, I agree this can be cleaned up in this particular case.

roccotripaldi replied:
We can deal with that in a separate PR. This one has lived too long!

mtias commented:
The approach here worries me a bit. If we start adding _ad hoc_ handling of sites (a central piece of how Calypso works) we are going to produce another monster like sites-list that we'll have a hard time testing and maintaining. There is a lot going on with sites, and many pieces rely on how it works, including routing. What happens if there's code that starts assuming your only site is the recently connected site because you have populated `state.sites` with a single item?

> I don't think it is a code smell, it's just that JPC requires more surgical handling of site's list. The behavior I'm proposing only needs to apply JPC process.

What's stopping other people from adding their own surgical handling of sites to accommodate a specific use case?

------

Could we take a moment and describe the main problem this is trying to solve? If it is speed, we need to make sites-list work faster regardless. The current plan is to take a look at what fields are being requested. If even with that optimization (which we haven't made yet so we don't know the impact) you still need a faster handling for connecting a site, then I suggest you convert the connection process for Jetpack to have nothing to do with `state.sites`. Example: create a specific section of the state tree to hold the connected site, do the single request, and then _wait_ for `QuerySites` to fetch again (or dispatch a fetch action) and re-populate `state.sites`.

The Calypso UI is not designed to have a single site but a multi-site user.

roccotripaldi commented:
> Could we take a moment and describe the main problem this is trying to solve?

Yes. We are trying to speed up Jetpack Connect specifically for folks with large site's lists.
Because right now, when you try to connect a Jetpack site, we must wait for `me/sites` to respond before we let you start the process.

And then once the site is connected, we wait for `me/sites` again before going to the final step.

With this PR, instead of waiting for `me/sites` in both of those cases, we wait only for the single site we care about. 

So at the begining of the process, we make a request only for that site to make sure it is not already in the site's list. And then after connecting, we request only the single site to add it to the list.

This works well, and makes the process much faster for folks with enormous site's lists. 

But since `me/sites` still runs regardless of the changes in this PR, there are cases where it will push the new site off the list. 

> The current plan is to take a look at what fields are being requested. If even with that optimization (which we haven't made yet so we don't know the impact

Never got that memo until today. :(

> then I suggest you convert the connection process for Jetpack to have nothing to do with state.sites. Example: create a specific section of the state tree to hold the connected site, do the single request, and then wait for QuerySites to fetch again (or dispatch a fetch action) and re-populate state.sites.

@tyxla - what do you think? let's abandon this and try with an other approach?

mtias commented:
> So at the begining of the process, we make a request only for that site to make sure it is not already in the site's list. And then after connecting, we request only the single site to add it to the list.

So it sounds you should optimize for the initial connect as a separate operation but, once it's connected, waiting for `/me/sites` is alright because that is needed in any case for Calypso to work properly.

roccotripaldi commented:
> So it sounds you should optimize for the initial connect as a separate operation but, once it's connected, waiting for /me/sites is alright because that is needed in any case for Calypso to work properly.

Good feedback, thanks for taking the time to look at this.

I think that since Calypso is already handling loading a full site lists, Jetpack connect doesn't need to at all. I think Jetpack should just deal with making the site in question instantly accessable after connection. 

This PR does that! It is blazing fast, and it uses Redux middleware in the exact way middleware was made for. I will say bye to the PR, but it is not easy. 🎵 🎻 🎻 🎵 🎻 🎻 🎵 

We will try to make the API a little faster, and try other approaches here in calypso.

tyxla commented:
You've both brought up some good points. Jetpack connection needs to be fast, while at the same time tempering with the sites state too much isn't recommended, as it could create flaws similar to what sites-list was doing.

So, what I like about this approach (the last approach in the PR) is that was quite to the point, and was using what we have right now, without altering it too much (like the first locking approach was doing).

> So it sounds you should optimize for the initial connect as a separate operation but, once it's connected, waiting for /me/sites is alright because that is needed in any case for Calypso to work properly.

This is where I disagree - if you see p7bbVw-1dO-p2 you will notice that waiting for `/me/sites` is not alright when the user has many sites (over 400 in that case). Jetpack Connect needs only the connecting site in order to function properly, but it still waits for all sites to be loaded. Personally I think it will be better if we load only the site we need so that Jetpack Connect can function properly, and then sites can load in the background, without the need to slow down the experience for the user. This is what I liked about this PR.

> If that is speed, we need to make /me/sites work faster regardless. The current plan is to take a look at what fields are being requested and streamline the request. If even with that optimization (which we haven't made yet so we don't know the impact) you still need a faster handling for connecting a site, then I suggest you convert the connection process for Jetpack to have nothing to do with state.sites. Example: create a specific section of the state tree to hold the connected site, do the single request, and then wait for QuerySites to fetch again (or dispatch a fetch action) and re-populate state.sites.

This is where I agree. If we don't want to overcomplexify sites more than it is right now, we'll probably need to handle this separately from sites. We're currently storing the connection/authorization data in a separate place from sites anyway. 

Let's try some alternative approaches without much tampering of the main sites tree. We might have to end up refactoring more of the JPC flows, but for the sake of it loading faster I think it'll totally be worth it.

mtias commented:
> Jetpack Connect needs only the connecting site in order to function properly, but it still waits for all sites to be loaded. Personally I think it will be better if we load only the site we need so that Jetpack Connect can function properly, and then sites can load in the background, without the need to slow down the experience for the user.

@tyxla that's what I meant. One thing is to do a single request to make connect work. Another is to use a single request to make a site work in Calypso in general (navigation, selectors, sidebar, etc). My point is you should only use a single request for making the connection process, but you should let QuerySites do its thing for actually using that site in Calypso, like all sites do.

The fact `/me/sites` is slow is an orthogonal problem that we are actively trying to improve.
